tests/client.test.js

```diff
@@ -231,6 +232,153 @@ tap.test('Flow Manager Client', async t => {
+        log,
+        {
+          kafkaConf,
+          components: {
+            commandsExecutor: {
+              consumerConf,
+              commandsTopic: topicsMap.cmd,
+              partitionsConsumedConcurrently: 1,
+            },
+          },
+        },
+      )
+      await client.start()
+
+      assert.teardown(async() => {
+        await client.stop()
+      })
+
+      let start1, end1, start2, end2
+      const executor1 = async() => {
+        start1 = Date.now()
+        await sleep(500)
+        end1 = Date.now()
+      }
+      const executor2 = async() => {
+        start2 = Date.now()
+        await sleep(500)
+        end2 = Date.now()
+      }
+
+      client.onCommand(command1, executor1)
+      client.onCommand(command2, executor2)
+
+      await commandIssuer.send({ topic: topicsMap.cmd, messages })
+
+      // wait that the messages have been received and processed
+      let whileCondition
+      do {
+        await sleep(300)
+        whileCondition = !(end1 && end2)
+      } while (whileCondition)
+
+      const consumerOffsets = await getConsumerOffsets(consumerGroup, topicsMap.cmd)
+      assert.equal(consumerOffsets.length, 2, 'Two partitions are available')
+
+      const firsExecutedEnd = start1 < start2 ? end1 : end2
+      const secondExecutedStart = start1 < start2 ? start2 : start1
+      const sequentialCondition = firsExecutedEnd < secondExecutedStart
+      assert.ok(sequentialCondition, 'Both commands have been executed sequentially')
+      assert.end()
+    })
+
+    t.test('execute commands in 2 partition concurrently if partitionsConsumedConcurrently is 2', async assert => {
+      const client = new FlowManagerClient(
+        log,
+        {
+          kafkaConf,
+          components: {
+            commandsExecutor: {
+              consumerConf,
+              commandsTopic: topicsMap.cmd,
+              partitionsConsumedConcurrently: 2,
+            },
+          },
+        },
+      )
+      await client.start()
+
+      assert.teardown(async() => {
+        await client.stop()
+      })
+
+      let start1, end1, start2, end2
+      const executor1 = async() => {
+        start1 = Date.now()
+        await sleep(500)
+        end1 = Date.now()
+      }
+      const executor2 = async() => {
+        start2 = Date.now()
+        await sleep(500)
+        end2 = Date.now()
+      }
+
+      client.onCommand(command1, executor1)
+      client.onCommand(command2, executor2)
+
+      await commandIssuer.send({ topic: topicsMap.cmd, messages })
+
+      // wait that the messages have been received and processed
+      let whileCondition
+      do {
+        await sleep(300)
+        whileCondition = !(end1 && end2)
+      } while (whileCondition)
+
+      const consumerOffsets = await getConsumerOffsets(consumerGroup, topicsMap.cmd)
+      assert.equal(consumerOffsets.length, 2, 'Two partitions are available')
+
+      const parallelCondition = start1 < end2 && start2 < end1
```

Comment:
Scriverei come commento la giustificazione di questa condizione che permette di dimostrare che la concorrenza funziona. La giustificazione è che essendo Node single threaded, comunque la condizione di await all'interno degli executor blocca l'esecuzione anche del secondo executor (o viceversa) anche se sono partiti insieme. Un altro modo in cui si potrebbe dimostrare questo è guardando il tempo totale di esecuzione. Nel caso di esecuzione sequenziale il tempo totale è almeno la somma dei due tempi di sleep, mentre in caso parallelo è ragionavole aspettarsi un tempo di esecuzione totale inferiore alla somme degli sleep.